source/common/json/json_sanitizer.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <string>
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Json {
+
+// Hand-rolled JSON sanitizer that has exactly the same behavior as serializing
+// through protobufs, but is more than 10x faster. From
+// test/common/json/json_sanitizer_speed_test.cc:
+//
+// ---------------------------------------------------------------------
+// Benchmark                           Time             CPU   Iterations
+// ---------------------------------------------------------------------
+// BM_ProtoEncoderNoEscape          1100 ns         1100 ns       636735
+// BM_JsonSanitizerNoEscape         18.1 ns         18.1 ns     38671848
+// BM_ProtoEncoderWithEscape        1376 ns         1375 ns       507137
+// BM_JsonSanitizerWithEscape       94.5 ns         94.5 ns      7402468
+//
+class JsonSanitizer {
+public:
+  // Constructing the sanitizer fills in a table with all escape-sequences,
+  // indexed by character. To make this perform well, you should instantiate the
+  // sanitizer in a context that lives across a large number of sanitizations.
+  JsonSanitizer();
+
+  /**
+   * Sanitizes a string so it is suitable for JSON. The buffer is
+   * used if any of the characters in str need to be escaped.
+   *
+   * @param buffer a string in which an escaped string can be written, if needed
+   * @param str the string to be translated
+   * @return the translated string_view.
+   */
+  absl::string_view sanitize(std::string& buffer, absl::string_view str) const;
+
+private:
+  // Character-indexed array of translation strings. If an entry is nullptr then
+  // the character does not require substitution. This strategy is dependent on
+  // the property of UTF-8 where all two-byte characters have the high-order bit
+  // set for both bytes, and don't require escaping for JSON. Thus we can
+  // consider each character in isolation for escaping.
+  std::unique_ptr<std::string> char_escapes_[256];
```

Comment:
32 bytes for every string and 8 bytes for every pointer, then here would be 10240 bytes external memory overhead?

Reply:
Most of the 256 element array has nullptr but I think your point is the footprint of this can be smaller. I think this could be an array of char[8], to allow fitting `\uABCD` (6 chars) and avoid the indirection. I'll make that change; then it's only 2k overhead, and will probably be faster too due to better cache behavior.

Reply:
Indeed this comment helped me to remove the level of indirection, making this code run significantly faster!

I could also just make this array 127 or 128 long, as no characters >= 127 get escaped, at the cost of one extra condition-check per character when doing sanitization. Let me try that.

Reply:
Making the array smaller and adding the extra conditional made things significantly slower so I won't commit that. 2k bytes overhead seems OK for something we only need one of.